clc/modules/object-storage/src/main/java/com/eucalyptus/objectstorage/ObjectStorageGateway.java

```diff
@@ -2013,113 +2055,317 @@ public DeleteBucketTaggingResponseType deleteBucketTagging(DeleteBucketTaggingTy
-    Bucket bucket = getBucketAndCheckAuthorization(request);
-    DeleteBucketCorsResponseType response = request.getReply();
+    String bucketName = "nullBucket";
+    DeleteBucketCorsResponseType response = null;
+
+    // Let most NullPointerExceptions be handled by the catch block.
     try {
+      Bucket bucket = getBucketAndCheckAuthorization(request);
+      bucketName = request.getBucket();
+      response = (DeleteBucketCorsResponseType) request.getReply();
+      
       BucketCorsManagers.getInstance().deleteCorsRules(bucket.getBucketUuid());
-    } catch (Exception e) {
-      InternalErrorException ex = new InternalErrorException(bucket.getBucketName() + "?cors");
-      ex.setMessage("An exception was caught while deleting the CORS configuration for bucket <" + bucket.getBucketName()+ ">, error: " + ex.getMessage());
-      throw ex;
+    } catch (S3Exception s3e) {
+      LOG.warn("Caught S3Exception while deleting the CORS configuration for bucket <" + 
+          bucketName + ">, CorrelationId: " + Contexts.lookup().getCorrelationId() + 
+          ", responding to client with: ", s3e);
+      throw s3e;
+    } catch (Exception ex) {
+      LOG.warn("Caught general exception while deleting the CORS configuration for bucket <" + 
+          bucketName + ">, CorrelationId: " + Contexts.lookup().getCorrelationId() + 
+          ", responding to client with 500 InternalError because of: ", ex);
+      throw new InternalErrorException(bucketName, ex);
     }
     return response;
   }
 
+  @Override
+  public PreflightCheckCorsResponseType preflightCors(PreflightCheckCorsType request) throws S3Exception {
+    String bucketName = "nullBucket";
+    PreflightCheckCorsResponseType response = null;
+
+    try {
+      // For a preflight request, we don't need to authenticate the client's 
+      // access to the bucket nor object in the request. We send back the 
+      // response headers regardless. This matches AWS behavior.
+      // We're not exposing any data from the bucket, just whether or
+      // not the client's proposed CORS request would be allowed or not,
+      // based on the CORS config. The actual CORS request can still fail
+      // based on authentication, independent of CORS.
+      // So, normally added to a method like this but not here:
+      // (left here commented out as an explainer)
+      // Bucket bucket = getBucketAndCheckAuthorization(request);
+      
+      bucketName = request.getBucket();
+      Bucket bucket = ensureBucketExists(bucketName);
+      String key = request.getKey();
+
+      PreflightRequest preflightRequest = request.getPreflightRequest();
+      response = request.getReply();
+
+      String requestOrigin = preflightRequest.getOrigin();
+      if (requestOrigin == null || requestOrigin.isEmpty()) {
+        CorsPreflightNoOriginException s3e = new CorsPreflightNoOriginException();
+        throw s3e;
+      }
```

Comment:
This looks like request validation, can it be moved up before the bucket lookup?